pkg/providers/cloudstack/apibuilder.go

```diff
@@ -0,0 +1,118 @@
+	return clusterapi.MachineDeployment(clusterSpec, workerNodeGroupConfig, kubeadmConfigTemplate, cloudstackMachineTemplate)
+}
+
+func MachineDeployments(clusterSpec *cluster.Spec, kubeadmConfigTemplates map[string]*bootstrapv1.KubeadmConfigTemplate, machineTemplates map[string]*cloudstackv1.CloudStackMachineTemplate) map[string]*clusterv1.MachineDeployment {
+	m := make(map[string]*clusterv1.MachineDeployment, len(clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
+
+	for _, workerNodeGroupConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		deployment := machineDeployment(clusterSpec, workerNodeGroupConfig,
+			kubeadmConfigTemplates[workerNodeGroupConfig.Name],
+			machineTemplates[workerNodeGroupConfig.Name],
+		)
+		m[workerNodeGroupConfig.Name] = &deployment
+	}
+	return m
+}
+
+func fillMachineTemplateAnnotations(machineConfig *v1alpha1.CloudStackMachineConfigSpec) map[string]string {
+	annotations := make(map[string]string, 0)
+	if machineConfig.DiskOffering != nil {
+		annotations[fmt.Sprintf("mountpath.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.MountPath
+		annotations[fmt.Sprintf("device.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Device
+		annotations[fmt.Sprintf("filesystem.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Filesystem
+		annotations[fmt.Sprintf("label.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Label
+	}
+
+	if machineConfig.Symlinks != nil {
+		links := make([]string, 0)
+		for key := range machineConfig.Symlinks {
+			links = append(links, fmt.Sprintf("%s:%s", key, machineConfig.Symlinks[key]))
+		}
+		// sorting for unit test determinism
+		sort.Strings(links)
+		annotations[fmt.Sprintf("symlinks.%s", constants.CloudstackAnnotationSuffix)] = strings.Join(links, ",")
+	}
+
+	return annotations
+}
+
+func setDiskOffering(machineConfig *v1alpha1.CloudStackMachineConfigSpec, template *cloudstackv1.CloudStackMachineTemplate) {
+	if machineConfig.DiskOffering == nil {
+		return
+	}
+
+	template.Spec.Spec.Spec.DiskOffering = cloudstackv1.CloudStackResourceDiskOffering{
+		CloudStackResourceIdentifier: cloudstackv1.CloudStackResourceIdentifier{
+			ID:   machineConfig.DiskOffering.Id,
+			Name: machineConfig.DiskOffering.Name,
+		},
+		CustomSize: machineConfig.DiskOffering.CustomSize,
+		MountPath:  machineConfig.DiskOffering.MountPath,
+		Device:     machineConfig.DiskOffering.Device,
+		Filesystem: machineConfig.DiskOffering.Filesystem,
+		Label:      machineConfig.DiskOffering.Label,
+	}
+}
+
+func CloudStackMachineTemplate(name string, machineConfig *v1alpha1.CloudStackMachineConfigSpec, now anywhereTypes.NowFunc) *cloudstackv1.CloudStackMachineTemplate {
```

Comment:
Maybe we should have the name being generated in this method?

Reply:
Name generation is complex, and depends on the usecase. It varies between the CLI and controller, and between old controller and new controller. I think it would be cleaner to leave that logic upstream and keep this method simple.

Reply:
I would expect the CLI and the new controller to use the same name generation logic. How do those two differ right now?

Reply:
When we generate the template name for cluster creation, we use `common.CPMachineTemplateName`. 

When we generate the template name for cluster Upgrade, we call `provider.getControlPlaneNameForCAPISpecUpgrade` which  checks if a new control plane template is required. If a new one is required by validating the previous EKS-A cluster spec, previous bundles, or AnyImmutableFieldChanged. If a new template is needed, it uses the same function as above, otherwise we use the existing one from the cp CloudStackMachineTemplate. 

The controller currently checks if any immutable field was changed on the CloudStackDatacenterConfig or CloudStackMachineConfigs and generates a new name if things have changed - otherwise it uses the one from the KCP in the cluster, at `cp.Spec.MachineTemplate.InfrastructureRef.Name`.

These are all sightly different ways to generate the name which make me nervous to just generate the name inside this function. It would also require feeding in a lot of context about the rest of the system to be able to check if a new one is needed or not.

Reply:
I understand that concern. But even if we don't generate the name in this method, is possible to have a shared codepath generate the name for both controller and CLI?

Reply:
I think sharing name generation logic makes sense, but I don't think it could be handled in this PR, which is intended to help make migration to cluster lifecycle api easier. I will create an issue for this